### ISSUE TITLE
[QUIC] How to handle graceful connection shutdown 

### ISSUE THREAD

**Issue by jkotalik** [dotnet/runtime#46542]
See https://github.com/dotnet/runtime/blob/master/src/libraries/System.Net.Quic/src/System/Net/Quic/Implementations/MsQuic/MsQuicConnection.cs#L370-L371

We'd like to be able to gracefully close an HTTP/3 connection without losing data. Ideally, if we call CloseAsync with a zero error code, either msquic or System.Net.Quic would wait for all streams to close.

**Comment by geoffkizer** [comment]
What's the underlying problem here? Do we not know how to make msquic do what we want? Trying to understand what the action here is...

**Comment by jkotalik** [comment]
The underlying problem we'd like to figure out is if there is a way to initiate a graceful connection closure with by calling CloseAsync on the QuicConnection. I'm not sure if there is a way to make msquic gracefully close a connection where it waits for all streams to close. That' probably a question for them.

**Comment by scalablecory** [comment]
Purely in terms of the protocol, there is no graceful connection shutdown. Closing a connection explicitly closes all streams without flushing. So, MsQuic is reflecting that. The correct way to manage this is to gracefully shutdown your streams individually and only close the connection once that is done. The way we initially implemented the API is just a little wrong so we were trying to do something we shouldn't have, I think, with a MsQuicStream that made this more natural flow impossible.

**Comment by scalablecory** [comment]
Assuming MsQuicStream's shutdown was fixed to work correctly (i.e. completion actually indicated a full graceful shutdown where all data has been sent and ACKed), would you still need this?

If so, can you go more into your scenario for how this would fit?

FWIW I believe the intended HTTP/3 workflow is to not explicitly close, but instead let a connection idle out. In this scenario the only reason you'd close the connection is for abortive purposes -- things like a protocol error or a heuristic trigger like too slow, no active streams for a period of time, client taking too long to finish up after a GOAWAY, etc. -- so graceful flushes aren't really needed.

**Comment by halter73** [comment]
> FWIW I believe the intended HTTP/3 workflow is to not explicitly close, but instead let a connection idle out.

Do you know of any documentation related to this workflow? With HTTP/3, is the idea that after all the streams are closed (including the control streams) both the client and the server will just stop using the QUIC connection? If I want the client to reconnect so it can pick up a DNS update or something, would closing all the streams be enough?

**Comment by karelz** [comment]
Triage: @scalablecory thinks this will be resolved with his PR.

**Comment by ManickaP** [comment]
Triage: this is not possible with quic, there's need to be a synchronization on an application level to know where the peer consumed all the data.

cc: @CarnaViire we had an email chain with discussion about this, could you summarize it here?

**Comment by ManickaP** [comment]
Actually now that we have stream counter as part of connection, we might be able to postpone [`ConnectionShutdown`](https://github.com/microsoft/msquic/blob/main/docs/api/ConnectionShutdown.md) until stream count is 0. 
If we'd decide to make this work, there are two options for this:
- `CloseAsync` doesn't complete until all streams are finished, `CancellationToken` might be used to abort waiting and send the shutdown anyway
- `CloseAsync` return immediately if there are any streams pending and remembers to call shutdown when they're all gone (same as what we do with handle release)

Notes:
- Is `errorCode == 0` enough for us to distinguish whether user wants graceful shutdown? The `errorCode` is application specific, 0 works for HTTP/3, but what about others?
- Or should `CloseAsync` be always graceful and count on cancellation token / Dispose being called from other thread? This might lead to deadlocked apps, awaiting `CloseAsync` till eternity.
- ATM we abort the connection in Dispose if it hasn't been shutdown yet, we should keep this, to be able to somehow kill the connection and not end up with memory leaks. But then we're having widely different contract then `Socket` where `Close` and `Dispose` are the same.

I'm not 100% sure this is worthwhile ATM, we should probably discuss this.

@jkotalik  @JamesNK: is anything in ASP.NET depending on `Connection.CloseAsync` to be graceful as stated in the issue? Do you have some hacks to work around this which you'd like get rid off? This info would helps us set a priority here.

**Comment by karelz** [comment]
@jkotalik @JamesNK any thoughts?

**Comment by JamesNK** [comment]
In ASP.NET Core we have two places where CloseAsync is called:

1. Transport abort - https://github.com/dotnet/aspnetcore/blob/ac748357e8dd2358217d5c69e08f404224f6985c/src/Servers/Kestrel/Transport.Quic/src/Internal/QuicConnectionContext.cs#L73-L88 - this should definitely send an immediate shutdown.
2. Transport dispose - https://github.com/dotnet/aspnetcore/blob/ac748357e8dd2358217d5c69e08f404224f6985c/src/Servers/Kestrel/Transport.Quic/src/Internal/QuicConnectionContext.cs#L52-L69 - I think this should also send an immediate shutdown. We track active HTTP/3 requests on a connection, and if the server has requested a connection close then the HTTP/3 connection will wait until requests are drained (Kestrel aborts the connection if not done after a timeout). After requests are drained or the timeout then the transport is disposed.

**Comment by ManickaP** [comment]
Not needed at the moment. Closing, we can revisit if any use case occurs in the future.
